### PR TITLE
sql: parallelize FK and UNIQUE constraints

### LINKED ISSUE #95184: Parallelize check constraints
**Is your feature request related to a problem? Please describe.**
Currently check constraints are executed serially. If an insert, for example, has more than one constraint to check, this causes the query to take longer in proportion to the number of constraints.

**Describe the solution you'd like**
Execute check constraints in parallel.

Jira issue: CRDB-23387

### PR DESCRIPTION
**Pull request by yuzefovich**
**rowcontainer: support concurrent reads and some cleanup**

This commit refactors the row container infrastructure in order to make
it possible to read from a single container concurrently (via separate
iterators). It required moving some of the scratch space from the
container directly into the iterators (so that the iterators don't share
anything between each other) as well as making a similar adjustment to
the `pebbleMapIterator`. As a result, it is now possible to safely
concurrently iterate over all types of row containers; however, the
iterators need to be created and closed under mutex protection (which
seems like an acceptable limitation). This commit also makes a minor
change to make the container a named parameter inside of the iterator
(which makes it easier to see when an iterator is touching the
container).

The ability to read from the same row container (after all rows have
been added to it) concurrently will be utilized by the follow-up commit
in order to support validation of multiple FK and UNIQUE checks in
parallel.

Additionally, this commit improves `RowIterator` interface a bit by
renaming `Row` to `EncRow` (since the method returns `EncDatumRow`s) and
introducing a new `Row` implementation which returns `tree.Datums`. The
new method is now utilized in the `sql.rowContainerIterator` which
removes some of the unnecessary conversions between different row types.

Release note: None

**sql: miscellaneous cleanup and improvement**

This commit performs a bunch of minor cleanups and improvements in
preparation for the follow-up commit, in which we parallelize the
execution of post-query checks. The main rationale for changes in this
commit is making it easier to reason about concurrency-safety of
DistSQL infra.

The following changes are made:
- remove unused receiver from `FinalizePlan` and
`finalizePlanWithRowCount` methods
- remove an unused parameter from `getPlanDistribution`
- remove the reference to `planner` from `getPlanDistribution` in favor
of passing a single boolean
- refactor `resetEvalCtx` to make it concurrency-safe by extracting out
the modification of `extraTxnState` into `resetPlanner`
- refactor how `topLevelQueryStats` are aggregated across all parts of
the stmt (i.e. across the main query, subqueries, and postqueries) to
make it easier to introduce concurrency-safety to it in a follow-up
- set `planFlagVectorized` only on the main query setup and explicitly
propagate whether the flow is vectorized into `saveFlows`
- fix a couple of typos in comments.

Release note: None

**sql: parallelize FK and UNIQUE constraints**

This commit makes it so that we parallelize the execution of multiple
"check plans" (which include FOREIGN KEY and UNIQUE constraint checks
that run as "postqueries"). The main idea is that we use the LeafTxns
(derived from the RootTxn that was used for the main mutation query) and
execute each check in a separate task. As a result, the checks should be
completed faster, especially so in multi-region environments.

This required introduction of mutex-protection for several
planning-infra-related objects:
- `saveFlows` function that populates miscellaneous info, needed for the
stmt bundle
- `associateNodeWithComponents` function that creates a mapping from
`planNode`s to DistSQL processors, needed for different EXPLAIN variants
- `topLevelQueryStats.add` function that aggregates some execution stats
across all "queries" of a stmt.

Additionally, this commit also needed to make `scanBufferNode`s safe for
concurrent use. The way this works is that in the main query we have
`bufferNode` which has already accumulated some rows into a row
container. That row container can be read from concurrently as long as
the separate iterators are created (the creation and the closure must be
mutex-protected though).

All of these things are synchronized via a single "global" mutex. We
could have introduced a separate mutex for each of these things, but
a single one seems acceptable given that these operations should be
pretty quick and occur at different points throughout the checks'
execution.

Fixes: #95184.

Release note (performance improvement): The execution of multiple
FOREIGN KEY and UNIQUE constraint checks has been parallelized in some
cases. As a result, these checks should be completed faster, especially
so in multi-region environments where the checks require cross-region
reads.

### CODE REVIEW

**Comment by rytaft** [comment]
Awesome! Do you think this will be backportable to 22.2?

**Review by cucaroach — APPROVED** [review]
:lgtm:

> Reviewed 20 of 20 files at r1, 6 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-:-NNN8e9D0csuUfnnbF_e:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @msirek, @rharding6373, and @yuzefovich)

___
*[`pkg/sql/distsql_running.go` line 1989 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNN27zI4j3FQTgcnLT2:-NNN27zI4j3FQTgcnLT3:b-prplb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5d7e725994ea3dd7580a4ae866520808f9d124d/pkg/sql/distsql_running.go#L1989)):*
> ```Go
> 	defer recv.stats.add(&postqueryRecv.stats)
> 	// TODO(yuzefovich): at the moment, errOnlyResultWriter is sufficient here,
> 	// but it may not be the case when we support cascades through the optimizer.
> ```

Curious why this comment is now obsolete.
___
*[`pkg/sql/distsql_running.go` line 445 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNN1H2R082TUfkTGf7M:-NNN1H2R082TUfkTGf7N:bc0rdxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d643d618e6b284d6084f8cf43dfa1a1a26474ec2/pkg/sql/distsql_running.go#L445)):*
> ```Go
> 		// Don't set the flag if we're running a parallel check for two reasons:
> 		// - it prevents a data race
> 		// - it is a postquery, and the flags are (mostly) about the main query.
> ```

What race does this prevent?  Also what are the ramifications of not setting the flag?
___
*[`pkg/sql/distsql_running.go` line 1941 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNN8Eyk2X6cf1msGWAy:-NNN8Eyk2X6cf1msGWAz:b-mndq1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/d643d618e6b284d6084f8cf43dfa1a1a26474ec2/pkg/sql/distsql_running.go#L1941)):*
> ```Go
> 	// have multiple checks to run.
> 	runParallelChecks := parallelizeChecks.Get(&dsp.st.SV) && len(plan.checkPlans) > 1
> 	if runParallelChecks {
> ```

Would there ever be a point where too many checks run in parallel would be a problem?   Like what if there were thousands of FK and uniqueness checks?   Is that even possible?  Could that overwhelm the system (well maybe not overhwhelm but be disruptive?).

**Review by yuzefovich — COMMENTED** [review]
> Do you think this will be backportable to 22.2?

Yes, I think so (disabled by default). We could probably backport this to 22.1 too if necessary since I don't think any of the code around this stuff has changed in the last couple of years.

Overall, this change is not without its risks since there could be other spots that I missed where mutex protection is needed. However, I've been stressing the test overnight on gceworker, and it didn't find any races.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-:-NNNJp2r1qN1Qbzffx-G:b5ro1ge)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @msirek, and @rharding6373)

___
*[`pkg/sql/distsql_running.go` line 1989 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNN27zI4j3FQTgcnLT2:-NNNDHXk1mIw4Q-Nr2vc:b-kjcdrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5d7e725994ea3dd7580a4ae866520808f9d124d/pkg/sql/distsql_running.go#L1989)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Curious why this comment is now obsolete.
</blockquote></details>

It was left at a point where we (myself and Radu) weren't sure exactly what infra would be needed to support cascades (this was like three years ago or something). This `errOnlyResultWriter` turned out to be sufficient, and we just never removed the TODO.
___
*[`pkg/sql/distsql_running.go` line 445 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNN1H2R082TUfkTGf7M:-NNNEWe71IXUCcWApieb:b-tcjkgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d643d618e6b284d6084f8cf43dfa1a1a26474ec2/pkg/sql/distsql_running.go#L445)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

What race does this prevent?  Also what are the ramifications of not setting the flag?
</blockquote></details>

In `PlanningCtx.getDefaultSaveFlowsFunc` we're checking this flag before storing some vectorized-engine-specific state for the stmt bundle. Now we can have one goroutine call `saveFlows` (which reads this flag) while another goroutine is setting the flag in `setupFlows`. Adding mutex synchronization between the two seems wrong, so I just removed the write part.

Generally speaking, the flags are set before we get to the execution engine, so the race is only possible for this particular `planFlagVectorized`.

Typing this out made me realize that we should just set it only on the main query path, fixed it in the first commit.
___
*[`pkg/sql/distsql_running.go` line 1941 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNN8Eyk2X6cf1msGWAy:-NNNCuLKBT50_VYpdI25:b-am9j6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d643d618e6b284d6084f8cf43dfa1a1a26474ec2/pkg/sql/distsql_running.go#L1941)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Would there ever be a point where too many checks run in parallel would be a problem?   Like what if there were thousands of FK and uniqueness checks?   Is that even possible?  Could that overwhelm the system (well maybe not overhwhelm but be disruptive?).
</blockquote></details>

Yes, that's a good point. I too realized that we should introduce a node-wide semaphore here, added. I also refactored the logic a bit to always run at least one check in the current goroutine (since otherwise this goroutine would just sit unused).

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96123)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @msirek, and @rharding6373)

___
*[`pkg/sql/distsql_running.go` line 445 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNN1H2R082TUfkTGf7M:-NNNQkSIEU46C1v6txbd:b-mq4vur) ([raw file](https://github.com/cockroachdb/cockroach/blob/d643d618e6b284d6084f8cf43dfa1a1a26474ec2/pkg/sql/distsql_running.go#L445)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

In `PlanningCtx.getDefaultSaveFlowsFunc` we're checking this flag before storing some vectorized-engine-specific state for the stmt bundle. Now we can have one goroutine call `saveFlows` (which reads this flag) while another goroutine is setting the flag in `setupFlows`. Adding mutex synchronization between the two seems wrong, so I just removed the write part.

Generally speaking, the flags are set before we get to the execution engine, so the race is only possible for this particular `planFlagVectorized`.

Typing this out made me realize that we should just set it only on the main query path, fixed it in the first commit.
</blockquote></details>

This flag is needed to be propagated into that `saveFlows` after all, so I changed it into an explicit boolean argument.

**Review by rharding6373 — CHANGES_REQUESTED** [review]
Very nice work! This came together quickly.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-:-NNOHhE06RGnbQ1Ns4ug:b-l75yrg)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @msirek, and @yuzefovich)

___
*[`pkg/sql/conn_executor.go` line 2881 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNNwNPo1baVbnPp33iX:-NNNwNPpBtUurg6LTdMH:b-u1mcis) ([raw file](https://github.com/cockroachdb/cockroach/blob/1203eb8afb7d8c7943e342a14e9cc15d943b36eb/pkg/sql/conn_executor.go#L2881)):*
> ```Go
> 	evalCtx.DescIDGenerator = ex.getDescIDGenerator()
> 
> 	// See resetPlanner for more context on this.
> ```

nit: s/this/setting the maximum timestamp for AOST read retries.
___
*[`pkg/sql/distsql_running.go` line 1942 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNOC89H8aF7oiNxb6aH:-NNOAisW5G7nZdxFOD2D:bt8w7bn) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a4529b6cc5ebd4609e492e73466e4ccd79f369/pkg/sql/distsql_running.go#L1942)):*
> ```Go
> 	runParallelChecks := parallelizeChecks.Get(&dsp.st.SV) &&
> 		len(plan.checkPlans) > 1 &&
> 		dsp.parallelChecksSem.ApproximateQuota() > 0
> ```

TIL about quota pools!
___
*[`pkg/sql/distsql_running.go` line 1944 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNO67KpAS2u43cm4lOd:-NNO67KpAS2u43cm4lOe:b-jd6dg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a4529b6cc5ebd4609e492e73466e4ccd79f369/pkg/sql/distsql_running.go#L1944)):*
> ```Go
> 		dsp.parallelChecksSem.ApproximateQuota() > 0
> 	if runParallelChecks {
> 		// At the moment, we rely on not using the newer DistSQL spec factory to
> ```

Why? If it's something we can probably do in the future, should we add a TODO here?
___
*[`pkg/sql/distsql_running.go` line 2177 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNOCF25448kGseuW-Tw:-NNOCF25448kGseuW-Tx:b-ufjiv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a4529b6cc5ebd4609e492e73466e4ccd79f369/pkg/sql/distsql_running.go#L2177)):*
> ```Go
> 		numParallelChecks = quota
> 	}
> 	for numParallelChecks > 1 {
> ```

I think this should be `for numParallelChecks > 0`. When it's 1, we will always have at least one parallel check if there are 2 or more `checkPlans` even if updating the quota pool fails, I think. The last time we can enter this loop is when `numParallelChecks` is 2, so if the last `TryAcquire` fails, `numParallelChecks` will be decremented to 1, then we end the loop. Then, below, we will always execute at least one `checkPlans`. It doesn't seem like the behavior we want, since we may execute a parallel check without successfully updating the semaphore.
___
*[`pkg/sql/distsql_running.go` line 2193 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNO7evB-4kq7Q7Cp_un:-NNO7evB-4kq7Q7Cp_uo:b-g6mpz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a4529b6cc5ebd4609e492e73466e4ccd79f369/pkg/sql/distsql_running.go#L2193)):*
> ```Go
> 	var wg sync.WaitGroup
> 	// Execute first numParallelChecks concurrently.
> 	for i := range checkPlans[:numParallelChecks] {
> ```

Could you add some logging for tracing purposes when the checks start and end? Something to the effect of "Executing N checks in parallel" should be sufficient for future debugging. Same with the remaining serial checks below.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96123)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @msirek, and @rharding6373)

___
*[`pkg/sql/distsql_running.go` line 1944 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNO67KpAS2u43cm4lOd:-NNOJ1zZ2mfUCJMdYKGx:b7cnvun) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a4529b6cc5ebd4609e492e73466e4ccd79f369/pkg/sql/distsql_running.go#L1944)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Why? If it's something we can probably do in the future, should we add a TODO here?
</blockquote></details>

This is due to the need to modify `scanBufferNode`s to make them concurrency-safe by walking the `planNode` tree. It shouldn't be a big lift to implement a walker over a plan produced by the DistSQL spec factory, but right now it doesn't seem worth it. Left a TODO.
___
*[`pkg/sql/distsql_running.go` line 2177 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNOCF25448kGseuW-Tw:-NNOJLTd7A6W7NgWKhYV:bydpzrz) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a4529b6cc5ebd4609e492e73466e4ccd79f369/pkg/sql/distsql_running.go#L2177)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

I think this should be `for numParallelChecks > 0`. When it's 1, we will always have at least one parallel check if there are 2 or more `checkPlans` even if updating the quota pool fails, I think. The last time we can enter this loop is when `numParallelChecks` is 2, so if the last `TryAcquire` fails, `numParallelChecks` will be decremented to 1, then we end the loop. Then, below, we will always execute at least one `checkPlans`. It doesn't seem like the behavior we want, since we may execute a parallel check without successfully updating the semaphore.
</blockquote></details>

Indeed, nice catch, thanks. Fixed.
___
*[`pkg/sql/distsql_running.go` line 2193 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNO7evB-4kq7Q7Cp_un:-NNOJSSL3CZ8HsQTufT2:bp5rrw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a4529b6cc5ebd4609e492e73466e4ccd79f369/pkg/sql/distsql_running.go#L2193)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you add some logging for tracing purposes when the checks start and end? Something to the effect of "Executing N checks in parallel" should be sufficient for future debugging. Same with the remaining serial checks below.
</blockquote></details>

Good point, added. I actually had this logging locally already but went on a related [detour](https://cockroachlabs.slack.com/archives/C01CNRP6TSN/p1675450766690309) :smiley:

**Review by rharding6373 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-:-NNONxl8BY_dO3UK1tSJ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @msirek)

___
*[`pkg/sql/distsql_running.go` line 2193 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-NNO7evB-4kq7Q7Cp_un:-NNOKxE46vKvLZb_MQKX:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/93a4529b6cc5ebd4609e492e73466e4ccd79f369/pkg/sql/distsql_running.go#L2193)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Good point, added. I actually had this logging locally already but went on a related [detour](https://cockroachlabs.slack.com/archives/C01CNRP6TSN/p1675450766690309) :smiley:  
</blockquote></details>

:thumbsup:

**Review by yuzefovich — COMMENTED** [review]
@msirek I see that you were typing something, so I'll wait for your feedback.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-:-NNbZvGa2HIjXZWdYieA:b4hzq0y)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @cucaroach, @msirek, and @rharding6373)

**Comment by yuzefovich** [comment]
> You can go ahead and merge. I was just wondering if using a buffered channel or some other method besides a mutex would reduce contention further. I profiled a test case and found we're spending a lot of time sleeping, so maybe there's a possibility here for future improvement

Good point. Your test case takes 4.8s w/ the parallelization enabled vs 4.3s w/o on my laptop (the row container spills to disk), if I increase `distsql_workmem` so that the row container doesn't spill, still there is a slow-down: 3.4s w/ vs 3.0s w/o. I think we'll need to make the row container concurrency-safe for reads (since we're done writing in the main query). I'll look into that.

**Comment by yuzefovich** [comment]
> I'll look into that.

I think this'll be actually pretty simple. On a quick WIP prototype is reduces the time for w/ parallelization from 4.8s to 2.4s (with spilling) and from 3.4s to 2.3s (no spilling). Will polish this tomorrow.

**Review by msirek — COMMENTED** [review]
On my M1 laptop, the test takes about 8.2s with the new version of the fix, and 7.9s on a commit before the fix.  Something must be different between your system and mine.

> Reviewed 20 of 20 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96123#-:-NNh45F05mnRUFZeIltV:b-tuhdtw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @cucaroach and @rharding6373)

**Comment by yuzefovich** [comment]
I added a temporary second commit here, but I think I'll extract it into a separate PR for ease of reviewing. Hold off on reviewing changes here.

**Comment by yuzefovich** [comment]
Alright, I cleaned up a change to the row containers to support concurrent reads and extracted it into a separate PR (#96836) for review (I'll still merge it as part of this PR eventually). With that change applied, in Mark's example I see the expected improvement locally when using demo (from 4.2s w/o parallelization to 2.4s w/ (w/ disk spilling) and from 2.5s w/o parallelization to 2.0s w/ (w/o disk spilling)).

**Comment by yuzefovich** [comment]
The first commit in this PR was approved in #96836 while the second and the third are effectively unchanged since I got the approval, so I'm going ahead and merging this.

Thanks for the reviews everyone!

bors r+

**Pull request merged**
